### DISCUSSION TITLE
Support docker-less installations (for Proxmox LXC)

### DISCUSSION THREAD

**Discussion by lordratner** [toverainc/willow-inference-server#65] [General]
Hi there,

I guess my first comment is that GitHub discussions aren't enabled on this repo, so I'm putting this here as an issue 🙃.

I'm going to run this on a Proxmox server, and because I want to share the GPU across multiple services, a VM won't work well. Instead I would like to use an LXC.

Problem is, docker doesn't play nice with LXCs. So if there is a way to do a "bare metal" install (or one specifically designed for an LXC if you have a lot of Proxmox users) instead of docker, it should help a lot for those running other services such as Frigate NVR, Plex/Emby/Jellyfin, Stable Diffusion, Compreface, etc since LXC can share a GPU.

I hope this is a possibility, but I'll confess I'm not advanced enough to know how to make it possible.

**Comment by kristiankielhofner** [comment]
Sorry about that, I've enabled discussions and moved this.

LXD supports container nesting. I use docker inside of containers all of the time, it's really under-appreciated!

I'm not sure if Proxmox exposes it but it's [one configuration parameter](https://bobcares.com/blog/lxd-and-docker-containers-nesting/) on the container.

**Reply by lordratner** [reply]
Yes, it is possible, however there are a couple issues.

The biggest is that unless you make the container privileged, which is not a recommended security practice, updates to Proxmox (especially kernel updates) can break docker LXCs.

Another issue is that Docker will try to use the VFS file system which can crush performance. I think they have finally enabled the nested ZFS filesystem support, but this is a new change.

I just went through this and the Proxmox team, which is very helpful, were adamant that they do not recommend or support using Docker within an LXC, only a VM.

So while it is possible (and I have a few I haven't been able to convert away from docker), it is not best practice.

But using a VM means the GPU cannot be shared with other services.

**Reply by lordratner** [reply]
Here's the note from the Proxmox documentation [Here:](https://pve.proxmox.com/wiki/Linux_Container)

`If you want to run application containers, for example, Docker images, it is recommended that you run them inside a Proxmox QEMU VM. This will give you all the advantages of application containerization, while also providing the benefits that VMs offer, such as strong isolation from the host and the ability to live-migrate, which otherwise isn’t possible with containers.`

And this helpful user gave me a list of threads where the Proxmox team actively discouraged the use of Docker in an LXC:

[My thread from broken Docker LXC](https://forum.proxmox.com/threads/updating-proxmox-breaks-docker-lxc.126720/#post-553701)

**Reply by kristiankielhofner** [reply]
I've been successfully running containers with docker (and GPU, etc) for years with nothing other than ```security.nesting: "true"``` set in the container configuration. Requiring setting ```security.privileged: "true"``` with ```security.nesting: "true"``` hasn't been a requirement in LXD for a long time. Per the [maintainer of LXD](https://discuss.linuxcontainers.org/t/check-privileged-or-unprivileged/238) the output from one of my containers with nesting:

```
ls -l /proc/cpuinfo
-r--r--r-- 1 nobody nogroup 69878 May 24 13:33 /proc/cpuinfo
```

If this is a limitation of Proxmox it should be taken up with them, of course.

In terms of running without Docker, there's no reason it fundamentally won't work but you will be in for a wild ride. Docker provides significant abstraction between CUDA kernel driver versions and the myriad of libraries inside the container up to and including specific versions of python frameworks pinned to CUDA, tensorrt, and other libraries that enable all of the functionality we require.

In terms of filesystem performance, other than the one-time build of the container WIS does exactly zero disk I/O after startup. It is a non-factor.

**Reply by lordratner** [reply]
Gotcha. I'm not sure why some LXC docker instances break and other don't. Hopefully Willow isn't one of them. I'll find out soon.

**Reply by midzelis** [reply]
Going on a tangent here, but chimming in... 

Aside: I've run proxmox for years( and ESXi before that, and FreeNAS before that)... and for years I was running docker inside LXC containers. But, there were always weirdness with it. Sometimes the docker would hose the networking, and it would even bring down the proxmox host. It would also cause the proxmox host to timeout during shutdown - which is not great when the UPS is trying to shut everything down.... I ended up creating a QEMU VM and putting docker in that. Its the best thing I ever did. No more reliability problems. Alpine in a VM start and stops as fast a LXC container. 

Plus, I actually have multiple proxmox hosts and with a VM I'm able to quickly move the VM from host to host (its on a shared NFS drive) simply by starting/stopping the VM from the host I want it to run on. All in all, moving docker to a VM is the best thing I did. 

My set up proxmox -> QEMU running alpine-server + docker -> and docker runs everything else. I also iSCSI mount the main /var/lib/docker drive, so the VM itself is mostly just a shell, and I can back up both the docker and the VM independently.

**Reply by lordratner** [reply]
Yeah, the VM solves many issues, but unfortunately it precludes me from sharing the GPU across multiple services, which is necessary on my setup. I don't want to put Jellyfin, Frigate, and Willow on the same VM, as that ruins a large benefit of virtualization.

Getting as many services to run "bare metal" in the LXC removes the issues with Docker. But some things will just require it.

**Reply by midzelis** [reply]
I haven't tried it, but do some research into vGPU unlock. Might help.

**Reply by tensiondriven** [reply]
I'm also running Proxmox, and have a main VM image which is running text-generation-webui across 2x3090 cards.  I have some available VRAM for Willow and would like to run Willow on this machine as well.

I would prefer to run willow outside of docker, along-side other processes that use the GPU's.  Eventually, I will likely want to run Willow under Unraid/Docker.

For now, I'll give it a shot with Docker on my main inference VM.  Most of my concern with wanting to skip Docker comes from overhead, not just in terms of CPU but the dependencies, libraries, disk space, abstractions; things that can go wrong, etc.  (This may be partially out of my own ignorance.)

So, I am +1 for the ability to run willow inference server outside of docker!

**Comment by tensiondriven** [comment]
PS Good luck with vGPU ... I was just happy to get passthrough working.  Here's a link to one of my favorite resources on vGPU under Proxmox:

https://www.youtube.com/watch?v=cPrOoeMxzu0

**Comment by tensiondriven** [comment]
Another reason that dockerless install is important to me: I think my needs are going to require some changes to willow, and it's important that I'm able to tinker with the code.  I know it's possible to shell in to a docker container, but unless things are mapped properly, I risk losing my work.

Making installation available without docker would remove this concern.

@kristiankielhofner I'm curious, do you and the other devs do development inside docker containers?  If so, maybe there are some simple precautions I could take.

**Reply by kristiankielhofner** [reply]
For Willow and WIS the respective `utils.sh` script in the project root has a `shell` option that maps everything for you. For development on both WIS and Willow we use a mixture of container/host depending on the specific task.
